frame/paged-list/src/paged_list.rs

```diff
@@ -0,0 +1,530 @@
+	}
+}
+
+impl<Prefix, Hasher, Value, ValuesPerPage>
+	StoragePagedListMeta<Prefix, Hasher, Value, ValuesPerPage>
+where
+	Prefix: StorageInstance,
+	Value: FullCodec,
+	Hasher: StorageHasher,
+	ValuesPerPage: Get<u32>,
+{
+	pub fn from_storage() -> Option<Self> {
+		let key = Self::key();
+
+		sp_io::storage::get(key.as_ref()).and_then(|raw| Self::decode(&mut &raw[..]).ok())
+	}
+
+	pub fn key() -> Hasher::Output {
+		(StoragePagedListPrefix::<Prefix>::final_prefix(), b"meta").using_encoded(Hasher::hash)
+	}
+
+	pub fn append_one<EncodeLikeValue>(&mut self, item: EncodeLikeValue)
+	where
+		EncodeLikeValue: EncodeLike<Value>,
+	{
+		// Note: we use >= here in case someone decreased it in a runtime upgrade.
+		if self.last_value >= ValuesPerPage::get() {
+			self.last_value = 0;
+			self.last_page.saturating_inc();
+		}
+		let key = page_key::<Prefix, Hasher>(self.last_page);
+		self.last_value.saturating_inc();
+		// Pages are `Vec` and therefore appendable.
+		sp_io::storage::append(key.as_ref(), item.encode());
+		self.store();
+	}
+
+	pub fn store(&self) {
+		let key = Self::key();
+		self.using_encoded(|encoded| sp_io::storage::set(key.as_ref(), encoded));
+	}
+
+	pub fn reset(&mut self) {
+		*self = Default::default();
+		Self::clear();
+	}
+
+	pub fn clear() {
+		let key = Self::key();
+		sp_io::storage::clear(key.as_ref());
+	}
+}
+
+pub struct Page<V> {
+	next: ValueIndex,
+	index: PageIndex,
+	// invariant: this is **never** empty
+	values: sp_std::vec::Vec<V>,
+}
+
+impl<V: FullCodec> Page<V> {
+	/// Decode a page with `index` from storage.
+	pub fn from_storage<Prefix: StorageInstance, Hasher: StorageHasher>(
+		index: PageIndex,
+	) -> Option<Self> {
+		let key = page_key::<Prefix, Hasher>(index);
+		let values = sp_io::storage::get(key.as_ref())
+			.and_then(|raw| sp_std::vec::Vec::<V>::decode(&mut &raw[..]).ok());
+		let values = values.unwrap_or_default();
+		if values.is_empty() {
+			None
+		} else {
+			Some(Self { next: 0, index, values })
+		}
+	}
+
+	/// Delete this page from storage.
+	pub fn delete<Prefix: StorageInstance, Hasher: StorageHasher>(&self) {
+		delete_page::<Prefix, Hasher>(self.index);
+	}
+}
+
+/// Delete a page with `index` from storage.
+// Does not live under `Page` since it does not require the `Value` generic.
+pub(crate) fn delete_page<Prefix: StorageInstance, Hasher: StorageHasher>(index: PageIndex) {
+	let key = page_key::<Prefix, Hasher>(index);
+	sp_io::storage::clear(key.as_ref());
+}
+
+/// Storage key of a page with `index`.
+// Does not live under `Page` since it does not require the `Value` generic.
+pub(crate) fn page_key<Prefix: StorageInstance, Hasher: StorageHasher>(
+	index: PageIndex,
+) -> Hasher::Output {
+	(StoragePagedListPrefix::<Prefix>::final_prefix(), b"page", index).using_encoded(Hasher::hash)
+}
+
+impl<V: Clone> Iterator for Page<V> {
+	type Item = V;
+
+	fn next(&mut self) -> Option<Self::Item> {
+		let val = self.values.get(self.next as usize).cloned();
```

Comment:
The cloning here seems a little unfortunate. Could you perhaps store `std::vec::IntoIter` instead of a `Vec` inside of `Page`? Then the `.clone` won't be necessary.

Reply:
Yea this is the only reason that we need to have the `Clone` bound on the `Value` generic, quite annoying indeed.  
I tried to store an interator before, but not sure if i did it correctly. Will try again.

Reply:
Nice, this made the code much easier and i could remove the `next` index, since the iterator tracks its own index.  
(and no `Clone`)